File: 1025-divisor-game/1025-divisor-game.java
class Solution {
    public boolean divisorGame(int n) {
        int[] dp = new int[n + 1];
        Arrays.fill(dp, -1);
        return helper(n,true,dp);
    }
    private boolean helper(int n, boolean turn,int[] dp){
        if(n == 2) return turn;
        else if(n < 2) return !turn;
        else if(dp[n] != -1) return (dp[n] == 1) ? turn : !turn;
        
        for(int x = 1; x < n; x++){
            if(n % x == 0 && helper(n - x,!turn,dp) == turn){
                 dp[n] = 1;
                 return turn;
            }
            
        }
        dp[n] = 0;   
        return !turn;
    }
    
}

/*
    Arrays.sort(intervals, (a,b)-> a[0] - b[0]); // increasing, based on st 
    


*/

File: 1025-divisor-game/1025-divisor-game.py
class Solution:
    def divisorGame(self, n: int) -> bool:
        
        @cache 
        def helper(n):
            if n == 1:
                return False 
            elif n == 2:
                return True 
            op1 = False 
            for i in range(1,n):
                if n % i == 0:
                    op1 = op1 or not helper(n - i)
            return op1 

        return helper(n)